### PR TITLE
Prac 7 code review request

### PR DESCRIPTION
**Pull request by chenhaoxiang-jcu**
Please review. @craigshoemark

### CODE REVIEW

File: prac_07/project_management.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+"""
+
+from prac_07.project import Project
+
+FILENAME = 'projects.txt'
+MENU = ('- (L)oad projects\n- (S)ave projects\n- (D)isplay projects\n- (F)ilter projects by date\n'
+        '- (A)dd new project\n- (U)pdate project\n- (Q)uit')
+
+
+def main():
```
Comment (craigshoemark) [inline prac_07/project_management.py:24]: Should we initialise projects list in main also?

File: prac_07/project_management.py [hunk 2]
```diff
@@ -45,3 +45,44 @@ def load_projects(filename):
...
+
+def load_new_projects():
+    filename = input("Filename to load: ")
+    projects = load_projects(filename)
+    return projects
+
+
+def proceed_saving_task(projects):
+    filename = input("Filename to save: ")
+    save_projects(projects, filename)
```
Comment (craigshoemark) [inline prac_07/project_management.py]: This function is essentially one input, could we get the input in main or even the save_projects function?

File: prac_07/project_management.py [hunk 3]
```diff
@@ -45,3 +45,44 @@ def load_projects(filename):
...
+        out_file.write("Name\tStart Date\tPriority\tCost Estimate\tCompletion Percentage")
+        for project in projects:
+            out_file.write(f"{project.name}\t{project.start_date}\t{project.priority}\t"
+                           f"{project.cost_estimate}\t{project.completion_percentage}\n")
+
+
+def load_new_projects():
+    filename = input("Filename to load: ")
+    projects = load_projects(filename)
+    return projects
```
Comment (craigshoemark) [inline prac_07/project_management.py]: Could this input be handled in main choice == 'L'  then parse filename in?

File: prac_07/project_management.py [hunk 4]
```diff
@@ -45,3 +45,44 @@ def load_projects(filename):
...
+
+
+def proceed_saving_task(projects):
+    filename = input("Filename to save: ")
+    save_projects(projects, filename)
+
+
+def display_projects(projects):
+    if not projects:
```
Comment (craigshoemark) [inline prac_07/project_management.py:97]: Nice!

File: prac_07/project_management.py [hunk 5]
```diff
@@ -45,3 +45,44 @@ def load_projects(filename):
...
+        print("No projects to display")
+    else:
+        completed_projects = []
+        incomplete_projects = []
```
Comment (craigshoemark) [inline prac_07/project_management.py:101]: Would this work if projects were sorted by priority?

File: prac_07/project_management.py [hunk 6]
```diff
@@ -45,3 +45,44 @@ def load_projects(filename):
...
+        for project in projects:
+            if project.is_completed():
+                completed_projects.append(project)
+            else:
+                incomplete_projects.append(project)
```
Comment (craigshoemark) [inline prac_07/project_management.py:106]: Could we use list comprehensions here ?

File: prac_07/project_management.py [hunk 7]
```diff
@@ -45,3 +45,44 @@ def load_projects(filename):
...
+        print("Incomplete projects: ")
+        for project in sorted(incomplete_projects):
+            print(f'  {project}')
```
Comment (craigshoemark) [inline prac_07/project_management.py]: Consistent spacing?

File: prac_07/project_management.py [hunk 8]
```diff
@@ -86,6 +86,30 @@ def display_projects(projects):
             print(f'  {project}')
 
 
+def add_project(projects):
+    print("Let's add a new project")
```
Comment (craigshoemark) [inline prac_07/project_management.py:161]: Nice!

File: prac_07/project_management.py [hunk 9]
```diff
@@ -86,6 +86,30 @@ def display_projects(projects):
...
+    name = input("Name: ")
+    start_date = input("Start date (dd/mm/yy): ")
+    priority = get_valid_number("Priority: ", int, 1, 999)
```
Comment (craigshoemark) [inline prac_07/project_management.py]: Should we call a function here?

File: prac_07/project_management.py [hunk 10]
```diff
@@ -86,6 +86,30 @@ def display_projects(projects):
...
+    cost_estimate = get_valid_number("Cost estimate: ", float, 0.0, 999999999.9)
```
Comment (craigshoemark) [inline prac_07/project_management.py]: What determines the price limit?

File: prac_07/project_management.py [hunk 11]
```diff
@@ -96,25 +103,25 @@ def display_projects(projects):
 
 def filter_projects(projects):
     date_string = get_valid_date("Show projects that start after date (dd/mm/yy): ")
-    date = datetime.datetime.strptime(date_string, '%d/%m/%Y').date()
+    date = datetime.datetime.strptime(date_string, DATE_FORMAT).date()
```
Comment (craigshoemark) [inline prac_07/project_management.py]: Nice work adding CONSTANT

File: prac_07/project_management.py [hunk 12]
```diff
@@ -125,9 +132,9 @@ def add_project(projects):
     print("Let's add a new project")
     name = input("Name: ")
     start_date = get_valid_date("Start date (dd/mm/yy): ")
-    priority = get_valid_number("Priority: ", int, 1, 999)
-    cost_estimate = get_valid_number("Cost estimate: $", float, 0.0, 999999999.9)
-    completion_percentage = get_valid_number("Percent complete: ", int, 0, 100)
+    priority = get_valid_number("Priority: ", int, LOWEST_PRIORITY, HIGHEST_PRIORITY)
+    cost_estimate = get_valid_number("Cost estimate: $", float, MINIMUM_COST, MAXIMUM_COST)
+    completion_percentage = get_valid_number("Percent complete: ", int, MINIMUM_PERCENTAGE, MAXIMUM_PERCENTAGE)
```
Comment (craigshoemark) [inline prac_07/project_management.py:166]: Nice CONSTANTS

File: prac_07/project.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
+class Project:
+    """Represent information about a project."""
+
+    def __init__(self, name='', start_date='', priority=0, cost_estimate=0.0, completion_percentage=0):
+        """Initialize a Project object."""
```
Comment (craigshoemark) [inline prac_07/project.py:5]: Nice docstrings

**Review by craigshoemark — COMMENTED** [review]
Good effort !!

File: prac_07/project_management.py (hunk 5 above)
Comment (chenhaoxiang-jcu) [reply prac_07/project_management.py:101]: I've already use __lt__ in Project class to compare priority and have sorted() below, but thanks.

File: prac_07/project_management.py (hunk 6 above)
Comment (chenhaoxiang-jcu) [reply prac_07/project_management.py:106]: Can list comprehension handle if, else? I'm not sure.

**Review by chenhaoxiang-jcu — COMMENTED** [review]
Hi, bro. Thanks a lot for reviewing. I think we might have different version of my code, cause I had changed some parts after I created a PR, but you are very welcome to try running my code in your computer😀. Thanks again!

**Pull request merged**